datastore/sqlite_datastore.nim

```diff
@@ -33,53 +39,149 @@ type
     containsStmt: ContainsStmt
     deleteStmt: DeleteStmt
     env: SQLite
+    getDataCol: BoundDataCol
     getStmt: GetStmt
     putStmt: PutStmt
     readOnly: bool
 
 const
-  IdType = "TEXT"
-  DataType = "BLOB"
-  TimestampType = "INTEGER"
-
   dbExt* = ".sqlite3"
-  tableTitle* = "Store"
+  tableName* = "Store"
+
+  idColName = "id"
+  dataColName = "data"
+  timestampColName = "timestamp"
+
+  idColIndex = 0
+  dataColIndex = 1
+  timestampColIndex = 2
+
+  idColType = "TEXT"
+  dataColType = "BLOB"
+  timestampColType = "INTEGER"
 
   # https://stackoverflow.com/a/9756276
   # EXISTS returns a boolean value represented by an integer:
   # https://sqlite.org/datatype3.html#boolean_datatype
   # https://sqlite.org/lang_expr.html#the_exists_operator
   containsStmtStr = """
     SELECT EXISTS(
-      SELECT 1 FROM """ & tableTitle & """
-      WHERE id = ?
+      SELECT 1 FROM """ & tableName & """
+      WHERE """ & idColName & """ = ?
     );
   """
 
   createStmtStr = """
-    CREATE TABLE IF NOT EXISTS """ & tableTitle & """ (
-      id """ & IdType & """ NOT NULL PRIMARY KEY,
-      data """ & DataType & """,
-      timestamp """ & TimestampType & """ NOT NULL
+    CREATE TABLE IF NOT EXISTS """ & tableName & """ (
+      """ & idColName & """ """ & idColType & """ NOT NULL PRIMARY KEY,
+      """ & dataColName & """ """ & dataColType & """,
+      """ & timestampColName & """ """ & timestampColType & """ NOT NULL
     ) WITHOUT ROWID;
   """
 
   deleteStmtStr = """
-    DELETE FROM """ & tableTitle & """
-    WHERE id = ?;
+    DELETE FROM """ & tableName & """
+    WHERE """ & idColName & """ = ?;
   """
 
   getStmtStr = """
-    SELECT data FROM """ & tableTitle & """
-    WHERE id = ?;
+    SELECT """ & dataColName & """ FROM """ & tableName & """
+    WHERE """ & idColName & """ = ?;
   """
 
   putStmtStr = """
-    REPLACE INTO """ & tableTitle & """ (
-      id, data, timestamp
+    REPLACE INTO """ & tableName & """ (
+      """ & idColName & """,
+      """ & dataColName & """,
+      """ & timestampColName & """
     ) VALUES (?, ?, ?);
   """
 
+template checkColMetadata(s: RawStmtPtr, i: int, expectedName: string) =
+  let
+    colName = sqlite3_column_origin_name(s, i.cint)
+
+  if colName.isNil:
+    raise (ref Defect)(msg: "no column exists for index " & $i)
+
+  if $colName != expectedName:
+    raise (ref Defect)(msg: "original column name for index " & $i & " was \"" &
+      $colName & "\" but expected \"" & expectedName & "\"")
```

Comment:
What do you mean by *"include the SQL statement text"*?

To see this error message in action, keep in mind this query:
```
SELECT timestamp as foo, id as baz, data as bar FROM ...
```
And change this line in the test:
https://github.com/status-im/nim-datastore/blob/06a94ec4cdebc794fed3ef1ad2603bc9edc21a25/tests/datastore/test_sqlite_datastore.nim#L147
So that it's `idCol = idCol(RawStmtPtr(prequery), 0)`
When the test is run a Defect will be raised with error message:
> Unhandled exception: original column name for index 0 was "timestamp" but expected "id" [Defect]

If the author of a custom query compares that kind of error message with the `SELECT` statement they wrote, I hope it would be clear that they mismatched the index with the column order they specified in `SELECT ...`.


However, I realize I am biased in this regard because I am thinking about it in terms of what led me to have this metadata check and error message in the first place.

Reply:
Noone executes isoltaed SQL statements. Instead, they run their complex program and when it bails with message `expected id`, they yet have to fugure out which program part did that. By adding this extra info, we can simplify the debugging process.

It's not mandatory, just an opportunity.

Reply:
Change made. I realized what you meant after your last reply in this thread. It's a very good idea, thanks!